oauth2client/contrib/django_util/__init__.py

```diff
@@ -265,43 +405,65 @@ class UserOAuth2(object):
 
     def __init__(self, request, scopes=None, return_url=None):
         """Initialize the Oauth2 Object
-        :param request: Django request object
-        :param scopes: Scopes desired for this OAuth2 flow
-        :param return_url: URL to return to after authorization is complete
-        :return:
+
+        Args:
+            request: Django request object
+            scopes: Scopes desired for this OAuth2 flow
+            return_url: The url to return to after the OAuth flow is complete,
+                 defaults to the request's current URL path
         """
         self.request = request
         self.return_url = return_url or request.get_full_path()
-        self.scopes = set(oauth2_settings.scopes)
+        self._scopes = set(oauth2_settings.scopes)
         if scopes:
-            self.scopes |= set(scopes)
-
-        # make sure previously requested custom scopes are maintained
-        # in future authorizations
-        credentials = storage.get_storage(self.request).get()
-        if credentials:
-            self.scopes |= credentials.scopes
+            self._scopes |= set(scopes)
 
     def get_authorize_redirect(self):
         """Creates a URl to start the OAuth2 authorization flow"""
         get_params = {
             'return_url': self.return_url,
-            'scopes': self.scopes
+            'scopes': self._get_scopes()
         }
 
-        return _redirect_with_params('google_oauth:authorize',
-                                     **get_params)
+        return _redirect_with_params('google_oauth:authorize', **get_params)
 
     def has_credentials(self):
         """Returns True if there are valid credentials for the current user
         and required scopes."""
-        return (self.credentials and not self.credentials.invalid and
-                self.credentials.has_scopes(self.scopes))
+        credentials = self._credentials()
+
+        x = (credentials and not credentials.invalid and
+                credentials.has_scopes(self._get_scopes()))
+        return x
+
+    def _credentials(self):
```

Comment:
Private instance method that does not alter the instance and only makes use of one attribute of `self` - these should always be pulled out of their classes and made module-scope module-private helper functions with their `self` parameter transformed into whatever was the one attribute of `self` that they actually needed.